lib/join/server.go

```diff
@@ -413,6 +448,9 @@ func makeHostCertsParams(
 	}
 
 	if authCtx.IsInstance {
+		if authCtx.Scope != scope {
+			return nil, trace.BadParameter("tried to re-join instance from scope %q into %q", authCtx.Scope, scope)
+		}
```

Comment:
This feels like a security check hidden inside makeHostCertsParams that doesn't really contribute to making the HostCertsParams. Putting it here also means it is checked last after all the more expensive join-method-specific tests. With a check like this can we try to do it as early as possible as soon as we have all the information necessary to do the check? Probably right next to the call to TokenAllowsRole